### PR TITLE
Refactor/tx signing

### PR DESCRIPTION
**Pull request by kyranjamie**
> [Download the latest build](https://github.com/blockstack/stacks-wallet/actions/runs/784668651)

This PR addresses some long-standing technical debt. Much code was repeated for each contract call when implementing delegated stacking. 
With these changes, a generic `<SignTransaction />` component is introduced. It is passed the options to sign a transaction. It handles UI depending according to whether it's a software or ledger wallet, and returns a signed transaction when necessary actions are complete.

This abstraction could be taken further to separate the UI from the logic to sign transactions. This is less urgent.

Further, a generic tx broadcast error is now used consistently across all transaction signing options.

- [x] Add password memory fix to refactored forms
- [x] Handle reject tx correctly

### CODE REVIEW

File: app/components/home/delegation-card.tsx [hunk 1]
```diff
@@ -5,19 +5,25 @@ import { useDispatch, useSelector } from 'react-redux';
...
 import { ErrorLabel } from '@components/error-label';
 import { ErrorText } from '@components/error-text';
+import { useMempool } from '@hooks/use-mempool';
 
 export const DelegationCard: FC = () => {
   const dispatch = useDispatch();
+  const { outboundMempoolTxs } = useMempool();
   const delegationStatus = useDelegationStatus();
-  const hasPendingRevokeCall = useSelector(selectHasPendingRevokingDelegationCall);
+  const poxInfo = useSelector(selectPoxInfo);
```
Comment (aulneau) [inline app/components/home/delegation-card.tsx:22]: could be own hook

File: app/components/home/stacking-promo-card.tsx [hunk 2]
```diff
@@ -4,13 +4,19 @@ import { useHistory } from 'react-router-dom';
...
+import { useMempool } from '@hooks/use-mempool';
+import { isDelegateStxTx } from '@utils/tx-utils';
+import { selectPoxInfo } from '@store/stacking';
 
 export const StackingPromoCard: FC = memo(() => {
   const history = useHistory();
-  const hasPendingDelegateStxCall = useSelector(selectHasPendingDelegateStxCall);
+  const { outboundMempoolTxs } = useMempool();
+  const poxInfo = useSelector(selectPoxInfo);
+  const hasPendingDelegateStxCall = outboundMempoolTxs.some(tx =>
```
Comment (aulneau) [inline app/components/home/stacking-promo-card.tsx:17]: this hasPending tx type could be its own hook that combines the mempool, pox hooks

File: app/components/tx-signing/sign-transaction-ledger.tsx [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+} from '../../modals/components/stacking-modal-layout';
+
+type SignTransactionLedgerProps = SignTransactionProps;
+
+export const SignTransactionLedger = (props: SignTransactionLedgerProps) => {
+  const { action, txOptions, isBroadcasting, onTransactionSigned, onClose } = props;
+
+  const { step: ledgerStep, isLocked } = usePrepareLedger();
+  const [hasSubmitted, setHasSubmitted] = useState(false);
+  // const [ledgerError, setLedgerError] = useState<string | null>(null);
```
Comment (aulneau) [inline app/components/tx-signing/sign-transaction-ledger.tsx:24]: remove

File: app/components/tx-signing/sign-transaction-software.tsx [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+import { isDecryptionError } from '@crypto/key-encryption';
+import { SignTransactionProps } from './sign-transaction';
+import { DecryptWalletForm } from '@modals/components/decrypt-wallet-form';
+import { createSoftwareWalletTokenTransferTx } from '@hooks/use-create-token-transfer-tx';
+import { capitalize } from '@utils/capitalize';
+import { blastUndoStackToRemovePasswordFromMemory } from '@utils/blast-undo-stack';
+
+type SignTransactionSoftwareProps = SignTransactionProps;
+
+export const SignTransactionSoftware = (props: SignTransactionSoftwareProps) => {
```
Comment (aulneau) [inline app/components/tx-signing/sign-transaction-software.tsx:21]:
```suggestion
export const SignTransactionSoftware: React.FC<SignTransactionSoftwareProps> = (props) => {
```

File: app/hooks/use-decrypt-wallet.ts [hunk 5]
```diff
@@ -15,14 +21,21 @@ export function useDecryptWallet() {
       if (!encryptedMnemonic || !salt) {
         throw new Error('`encryptedMnemonic` or `salt` undefined');
       }
```
Comment (aulneau) [inline app/hooks/use-decrypt-wallet.ts:23]:
```suggestion
      if (!encryptedMnemonic) throw new Error('`encryptedMnemonic` undefined');
      if (!salt) throw new Error('`salt` undefined');
```

Comment (aulneau) [reply app/hooks/use-decrypt-wallet.ts:23]: for DX

File: app/modals/delegated-stacking/delegated-stacking-modal.tsx [hunk 6]
```diff
@@ -1,50 +1,41 @@
 import React, { FC, useState, useCallback } from 'react';
-import { useSelector, useDispatch } from 'react-redux';
-import { Modal } from '@stacks/ui';
+import { useSelector } from 'react-redux';
 import { useHistory } from 'react-router-dom';
-import { MempoolTransaction } from '@blockstack/stacks-blockchain-api-types';
 import { useHotkeys } from 'react-hotkeys-hook';
 import { BigNumber } from 'bignumber.js';
+import { Modal } from '@stacks/ui';
```
Comment (aulneau) [inline app/modals/delegated-stacking/delegated-stacking-modal.tsx:6]: should use the new modal i added from the ui updates

File: app/modals/delegated-stacking/delegated-stacking-modal.tsx [hunk 7]
```diff
@@ -55,190 +46,55 @@ interface StackingModalProps {
...
-  const { createSoftwareTx } = useCreateSoftwareTx();
-  const { step: ledgerStep, isLocked } = usePrepareLedger();
+  const { broadcastTx, isBroadcasting } = useBroadcastTx();
+
+  const { refetch } = useMempool();
 
   const api = useApi();
 
-  const { poxInfo, balance } = useSelector((state: RootState) => ({
+  const { poxInfo } = useSelector((state: RootState) => ({
```
Comment (aulneau) [inline app/modals/delegated-stacking/delegated-stacking-modal.tsx:59]: can be `usePoxInfo` hook

File: app/modals/delegated-stacking/delegated-stacking-modal.tsx [hunk 8]
```diff
@@ -55,190 +46,55 @@ interface StackingModalProps {
...
+      },
+      tx: signedTx,
+    });
 
   return (
-    <Modal isOpen headerComponent={header} footerComponent={footer} {...modalStyle}>
-      {body}
-    </Modal>
+    <TxSigningModal
+      action="initate delegation"
```
Comment (aulneau) [inline app/modals/delegated-stacking/delegated-stacking-modal.tsx:91]: should action be enum?

File: app/modals/revoke-delegation/revoke-delegation-modal.tsx [hunk 9]
```diff
@@ -1,238 +1,66 @@
...
+          await safeAwait(watchForNewTxToAppear({ txId, nodeUrl: api.baseUrl }));
+          await refetchMempool();
+          dispatch(homeActions.closeRevokeDelegationModal());
+        },
+        onFail: (error?: any) => {
+          if (error) setNodeResponseError(error);
+        },
+        tx: signedTx,
+      }),
+    [api.baseUrl, broadcastTx, dispatch, refetchMempool]
```
Comment (aulneau) [inline app/modals/revoke-delegation/revoke-delegation-modal.tsx:52]: should `watchForNewTxToAppear`, `homeActions` be included in deps? seems like there are a few missing, but might be wrong if the eslint rule hasn't yelled

File: app/modals/revoke-delegation/revoke-delegation-modal.tsx [hunk 10]
```diff
@@ -1,238 +1,66 @@
...
+  );
 
   return (
-    <Modal isOpen headerComponent={header} footerComponent={footer} {...modalStyle}>
-      {body}
-    </Modal>
+    <TxSigningModal
+      action="revoke delegation"
```
Comment (aulneau) [inline app/modals/revoke-delegation/revoke-delegation-modal.tsx:57]: enum?

File: app/modals/send-stx/send-stx-modal.tsx [hunk 11]
```diff
@@ -0,0 +1,278 @@
...
+
+  const [nodeResponseError, setNodeResponseError] = useState<PostCoreNodeTransactionsError | null>(
+    null
+  );
+  const [loading, setLoading] = useState(false);
+  const interactedWithSendAllBtn = useRef(false);
+  const { nonce } = useLatestNonce();
+
+  const sendStx = (tx: StacksTransaction) => {
+    console.log(tx);
```
Comment (aulneau) [inline app/modals/send-stx/send-stx-modal.tsx:76]: remove

File: app/modals/send-stx/send-stx-modal.tsx [hunk 12]
```diff
@@ -0,0 +1,278 @@
...
+    memo: form.values.memo,
+    nonce: new BN(nonce),
+  };
+
+  const [calculatingMaxSpend, setCalculatingMaxSpend] = useState(false);
+
+  const closeModal = () => dispatch(homeActions.closeTxModal());
+
+  const updateAmountFieldToMaxBalance = async () => {
+    interactedWithSendAllBtn.current = true;
```
Comment (aulneau) [inline app/modals/send-stx/send-stx-modal.tsx:179]: curious, why are you using a ref here?

File: app/modals/stacking/stacking-modal.tsx [hunk 13]
```diff
@@ -58,36 +33,19 @@ export const StackingModal: FC<StackingModalProps> = props => {
...
-    })
-  );
-
-  const initialStep =
-    walletType === 'software'
-      ? StackingModalStep.DecryptWalletAndSend
-      : StackingModalStep.SignWithLedgerAndSend;
-
-  const [step, setStep] = useState(initialStep);
+  const { poxInfo, coreNodeInfo } = useSelector((state: RootState) => ({
```
Comment (aulneau) [inline app/modals/stacking/stacking-modal.tsx:41]: could be own hook

**Review by aulneau — APPROVED** [review]

File: app/modals/revoke-delegation/revoke-delegation-modal.tsx (hunk 10 above)
Comment (kyranjamie) [reply app/modals/revoke-delegation/revoke-delegation-modal.tsx:57]: I don't see why, this is really just content.

Comment (aulneau) [reply app/modals/revoke-delegation/revoke-delegation-modal.tsx:57]: ah ok, makes sense

File: app/modals/send-stx/send-stx-modal.tsx (hunk 12 above)
Comment (kyranjamie) [reply app/modals/send-stx/send-stx-modal.tsx:179]: I think to not cause a rerender. I suppose this could also be a `let` outside the component, too.

Comment (aulneau) [reply app/modals/send-stx/send-stx-modal.tsx:179]: ah no, ref makes sense then

File: app/modals/revoke-delegation/revoke-delegation-modal.tsx (hunk 9 above)
Comment (kyranjamie) [reply app/modals/revoke-delegation/revoke-delegation-modal.tsx:52]: Both of these are es imports, their reference will never change, so afaik don't need to be listed as deps.

File: app/modals/stacking/stacking-modal.tsx (hunk 13 above)
Comment (kyranjamie) [reply app/modals/stacking/stacking-modal.tsx:41]: opened refactor issue for hooks

File: app/modals/revoke-delegation/revoke-delegation-modal.tsx [hunk 14]
```diff
@@ -1,238 +1,66 @@
...
+      txDetails={getRevocationTxOptions()}
```
[the remark this answers is not in the document]
Comment (kyranjamie) [reply app/modals/revoke-delegation/revoke-delegation-modal.tsx:58]: the fn is usecalled'd? Isn't this sufficient?

**Pull request merged**